KEH-1307 - Fix Major Linting Warnings

sebtheo opened the pull request:
## What type of PR is this?

- [x] Refactor
- [ ] Feature
- [x] Bug Fix
- [x] Optimization
- [ ] Documentation Update

### What changed?

#### Main Changes

The `/frontend/.eslintrc.cjs` would turn off a lot of rules and would let a lot of stuff through. I turned those off and fixed all these errors.

This PR involves a lot of refactoring, mainly moving stuff to prevent init errors and introducing move useCallback/ref/memo.

#### Secondary Changes

- `backend/src/config/rateLimiter.js` introduces increase to rate limit when running NODE_ENV as development. I changed this because when the frontend would occasionally decide to spam the backend, it would cause a rate limit.
- `frontend/.eslintrc.cjs` rules have been reduced to just one.
- `import React from 'react'` was in practically every frontend file, which is not needed.

### Documentation?

I don't think so.

### How to review?

Mark out a couple hours to painfully look through the code.

sebtheo committed:
chore(rateLimiter): increase number of requests per minute for testing

increased this as testing is tough when it's 1 per second

sebtheo committed:
fix(lint): App.js

sebtheo committed:
refactor(TechManage): fix lint errors

sebtheo committed:
fix(React): no-unused-vars flag

With React 17+, importing React is no longer necessary unless we use React.function like React.createElement .

sebtheo committed:
fix(lint): no-unused-var

sebtheo committed:
fix(ReviewPage): lint

sebtheo committed:
fix(StatisticsPage): lint

sebtheo committed:
fix(CopilotPage): lint

sebtheo committed:
fix(lint): no-unused-vars

sebtheo committed:
fix(CopilotPage): lint

sebtheo committed:
fix(RadarPage: no-unused-var

sebtheo committed:
fix(RadarPage): fix warnings

sebtheo committed:
ci(eslint): change frontend rules

sebtheo committed:
chore: lint pages

sebtheo committed:
chore(draggingQuadrant): fix states for quadrant positions

sebtheo committed:
merge 'main' into this branch

sebtheo committed:
Merge branch 'main' into KEH-1307-fix-codeql-warnings

sebtheo committed:
fix(lint): no-unused-vars

sebtheo committed:
ci(eslint): update config

sebtheo committed:
ci(lint): set-state-in-effect

sebtheo committed:
ci(lint): exhaustive deps

sebtheo committed:
ci(lint): no-unused-vars

sebtheo committed:
ci(lint): set-state-in-effect

sebtheo committed:
ci(CopilotPage): lint

sebtheo committed:
fix(Header): lint set-state-in-effect

sebtheo committed:
fix(HelpModal): lint set-state-in-effect

sebtheo committed:
fix(ReviewPage): lint format

sebtheo committed:
fix(eslint): rm old config

TotalDwarf03 reviewed (commented):
The changes look ok but this PR breaks the application in a few areas.

Lots of problems:

- Tech Radar, Stats and Project pages fail to render due to too many redirects.
- The Review page has a `Maximum update depth exceeded` error.
- The admin page is constantly pinging the backend to check for authentication.
- The tabs on the copilot section are also broken. When clicked, they just redirect to the same page.

The only page I can use without error with this PR is the homepage.

I can call to show the various errors if needed. We shouldn't merge this as it makes the tool unusable

sebtheo committed:
chore: rm console.log

sebtheo commented on frontend/src/components/Banner/Banner.js:
```diff
@@ -20,13 +20,11 @@ const Banner = ({ title, description, type = 'info', onClose }) => {
-      setIsVisible(false);
-    }
-  }, [bannerId]);
+  // Check localStorage to see if this banner was previously dismissed
```
It used to but seems it's not.

sebtheo commented on frontend/src/components/Admin/TechManage.js:
```diff
@@ -118,7 +119,48 @@ const TechManage = () => {
+      }
+
+      const data = await response.json();
+      console.log(data);
```
Removed.

sebtheo committed:
fix(Banner): incorrect dismissal logic

sebtheo committed:
Merge branch 'main' of https://github.com/ONS-Innovation/keh-digital-landscape into KEH-1307-fix-codeql-warnings

sebtheo committed:
chore(RadarPage): fix post merge

sebtheo committed:
chore(Projects): fix post merge

sebtheo committed:
fix(ReviewPage): Max depth exceeded

sebtheo committed:
chore: add name to search inputs

sebtheo committed:
chore(Projects): lint

sebtheo committed:
fix(CopilotPage): unable to switch tabs

sebtheo committed:
Merge branch 'main' into KEH-1307-fix-codeql-warnings

sebtheo committed:
fix(copilot): /token requested 5 times

sebtheo committed:
fix(allTechColumns): add Architectures

sebtheo committed:
chore: format `make help`

sebtheo committed:
test: add 500ms timeout

sebtheo committed:
test: rm useless function

sebtheo committed:
test(techreviewer): increase precision of test

sebtheo committed:
test(techradar): increase precision of test

sebtheo committed:
fix(CopilotPage): duplicate redundant requests

The page for /copilot/team/{team} would send a tonne of requests due to incorrect state usage. I've reordered some logic and introduced useRefs to ensure the requests are only sent when needed.

TotalDwarf03 reviewed (approved):
Looks good. Works locally. Good work 👍

mwirikia committed:
Merge branch 'main' into KEH-1307-fix-codeql-warnings
